implementations/rust/ockam/ockam/src/entity/profile/profile_helpers.rs

```diff
@@ -0,0 +1,144 @@
+// FIXME: Give me better name, pls
+
+use crate::entity::profile::event_identifier::EventIdentifier;
+use crate::entity::profile::profile_change::{ProfileChange, ProfileKeyPurpose, ProfileKeyType};
+use crate::entity::profile::profile_change_type::ProfileChangeType::{CreateKey, RotateKey};
+use crate::entity::profile::proof::{ProfileChangeProof, Signature, SignatureType};
+use crate::entity::{Profile, ProfileChangeEvent};
+use crate::OckamError;
+use ockam_vault_core::secret::Secret;
+
+impl Profile {
+    pub(crate) fn get_last_event_id(&self) -> Result<EventIdentifier, ockam_core::Error> {
+        if let Some(e) = self.change_history.as_ref().last() {
+            Ok(e.identifier().clone())
+        } else {
+            Err(OckamError::InvalidInternalState.into())
+        }
+    }
+
+    pub(crate) fn find_key_change_in_event<'a>(
+        event: &'a ProfileChangeEvent,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Option<&'a ProfileChange> {
+        event
+            .changes()
+            .as_ref()
+            .iter()
+            .rev()
+            .find(|c| match c.change_type() {
+                CreateKey(change) => {
+                    change.key_type() == key_type
+                        && change.key_purpose() == key_purpose
+                        && change.label() == label
+                }
+                RotateKey(change) => {
+                    change.key_type() == key_type
+                        && change.key_purpose() == key_purpose
+                        && change.label() == label
+                } // RevokeKey(event) => {
+                  //     event.key_type() == key_type && event.key_purpose() == key_purpose && event.label() == label
+                  // }
+            })
+    }
+
+    pub(crate) fn find_last_key_event(
+        &self,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Result<&ProfileChangeEvent, ockam_core::Error> {
+        self.change_history
+            .as_ref()
+            .iter()
+            .rev()
+            .find(|e| Self::find_key_change_in_event(e, label, key_type, key_purpose).is_some())
+            .ok_or(OckamError::InvalidInternalState.into())
+    }
+
+    pub(crate) fn find_key_event_before(
```

Comment:
I don't see any reason why we could just expose an array ref and let the caller do their own searching.

Reply:
We can, but this code is used from other places (verify function) + I think moving this code to separate function makes caller side look cleaner + there is probability we'll be reusing this function.